nf_core/components/update.py

```diff
@@ -0,0 +1,851 @@
+                    component, self.component_type, modules_repo=modules_repo, installed_sha=current_version
+                )
+            else:
+                version = modules_repo.get_latest_component_version(component, self.component_type)
+
+            if current_version is not None and not self.force:
+                if current_version == version:
+                    if self.sha or self.prompt:
+                        log.info(f"'{component_fullname}' is already installed at {version}")
+                    else:
+                        log.info(f"'{component_fullname}' is already up to date")
+                    continue
+
+            # Download component files
+            if not self.install_component_files(component, version, modules_repo, install_tmp_dir):
+                exit_value = False
+                continue
+
+            if patch_relpath is not None:
+                patch_successful = self.try_apply_patch(
+                    component,
+                    modules_repo.repo_path,
+                    patch_relpath,
+                    component_dir,
+                    component_install_dir,
+                    write_file=False,
+                )
+                if patch_successful:
+                    log.info(f"{self.component_type[:-1].title()} '{component_fullname}' patched successfully")
+                else:
+                    log.warning(
+                        f"Failed to patch {self.component_type[:-1]} '{component_fullname}'. Will proceed with unpatched files."
+                    )
+                all_patches_successful &= patch_successful
+
+            if dry_run:
+                if patch_relpath is not None:
+                    if patch_successful:
+                        log.info("Current installation is compared against patched version in remote.")
+                    else:
+                        log.warning("Current installation is compared against unpatched version in remote.")
+                # Compute the diffs for the component
+                if self.save_diff_fn:
+                    log.info(
+                        f"Writing diff file for {self.component_type[:-1]} '{component_fullname}' to '{self.save_diff_fn}'"
+                    )
+                    try:
+                        ModulesDiffer.write_diff_file(
+                            self.save_diff_fn,
+                            component,
+                            modules_repo.repo_path,
+                            component_dir,
+                            component_install_dir,
+                            current_version,
+                            version,
+                            dsp_from_dir=component_dir,
+                            dsp_to_dir=component_dir,
+                        )
+                        updated.append(component)
+                    except UserWarning as e:
+                        if str(e) != "Module is unchanged":
```

Comment:
I had to revert this because the error message is part of `ModulesDiffer` which has only helper functions, doesn't have access to self. We can refactor this whole class in a different PR.